### PR TITLE
feat(Core/CreatureAddon): increased visibility for large creatures

### PR DESCRIPTION
**Pull request by Stoabrogga**
##### CHANGES PROPOSED:
- Add a new column "isLarge" to `creature_addon` and `creature_template_addon`, holding a boolean value controlling if the creature is large or not (default: 0).
- Large creatures have their visibility increased to MAX_VISIBILITY_DISTANCE (250.0f).
- Large creatures are updated even if they are out of normal range of a player. I had to increase the cell visits for the players during map updates, but only trigger updates for large creatures, so it should not cost much performance. The large creatures are not active, they will stop getting updates after the player moves too far away.
- I tried to keep the code changes to a minimum in order to not break existing behaviour.
- I ported all large creatures from the TC database to AC. Examples:
  - Doomwalker (also added his missing waypoint movement)
  - Fel Reaver Sentry: Located in Abyssal Shelf, they are only used for the bombing runs
  - Dun Niffelem Spear Chain Bunny (Phase 2): They are used to cast the chains holding Hodir's Spear in Dun Niffelem (they were barely visible before)

Example: Doomwalker before / after:
WoWScrnShot_092319_170850
WoWScrnShot_092319_170958

##### ISSUES ADDRESSED:
none

##### TESTS PERFORMED:
- tested build on Ubuntu 16.04 / clang 7
- tested successfully in-game

##### HOW TO TEST THE CHANGES:
- Doomwalker: `.go -3465.99 312.405 93.4952 530`
- Fel Reaver Sentry: `.go 714.31 1885.59 176.008 530`
- Hodir's Spear: `.go 7235.8 -2844.83 822.885 571`

##### KNOWN ISSUES AND TODO LIST:
none

##### Target branch(es):
- [x] Master
 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

File: data/sql/updates/pending_db_world/rev_1569250786242020759.sql [hunk 1]
```diff
@@ -0,0 +1,14 @@
+INSERT INTO `version_db_world` (`sql_rev`) VALUES ('1569250786242020759');
+
+ALTER TABLE `creature_addon` ADD COLUMN `isLarge` TINYINT(3) UNSIGNED NOT NULL DEFAULT 0 AFTER `emote`;
+ALTER TABLE `creature_template_addon` ADD COLUMN `isLarge` TINYINT(3) UNSIGNED NOT NULL DEFAULT 0 AFTER `emote`;
```
Comment (FrancescoBorzi) [inline data/sql/updates/pending_db_world/rev_1569250786242020759.sql]: why tinyint? It can be boolean

Comment (Stoabrogga) [reply data/sql/updates/pending_db_world/rev_1569250786242020759.sql]: There is no built-in data type "boolean", it is a synonym for tinyint(1). But you're right, tinyint(1) suffices here. I have pushed a new commit.

**Comment by pklloveyou** [comment]
UPDATE `creature_addon` SET `isLarge` = 1 WHERE `guid` IN (67001);

**Comment by FrancescoBorzi** [comment]
@pklloveyou  just a suggestion... better formatting when posting comments with SQL code:
 
image

**Comment by pklloveyou** [comment]
```sql
UPDATE `creature_addon` SET `isLarge` = 1 WHERE `guid` IN (67001,118191,98159,103290,112693,51838);
```

**Comment by Stoabrogga** [comment]
@pklloveyou 
I used another approach now: I ported all large creatures from the TC database to AC. This also includes the ones you mentioned with the exception of Setis, which I added manually.

Could you please test this? Not all creatures, as there are thousands now set to "large", but perhaps a few of them and generally explore the world a bit.

**Comment by pklloveyou** [comment]
I did a few random tests and found that some of the creatures didn't exist.,

**Comment by FrancescoBorzi** [comment]
@pklloveyou can you mention which creatures didn't exist and their IDs / GUIDs ?

**Comment by pklloveyou** [comment]
Like ID90.  Too many..Is there a way to distinguish in bulk

**Comment by pklloveyou** [comment]
Is there a way to make a lot of sense of distinction?

**Comment by pklloveyou** [comment]
One by one,I feel like I'm going crazy.

**Comment by FrancescoBorzi** [comment]
just the ones you tested, of course we cannot test all of them manually because there are too many

**Comment by Stoabrogga** [comment]
Creature ID 90 exists in my DB, it is the Sea Giant. It is very important to use a fresh AC DB here, not a modified DB.

**Comment by pklloveyou** [comment]
I was wrong,the creature exists, if not present the server will display an error at startup

**Comment by FrancescoBorzi** [comment]
@pklloveyou a generall tip, to easily view/edit your DB you can use: https://github.com/azerothcore/Keira3/releases

**Comment by pklloveyou** [comment]
creature_addon 112693 It's not here.

**Comment by pklloveyou** [comment]
51838,152166,152165,247237,132193,132194 creature_addon

**Comment by pklloveyou** [comment]
image
China's international network sucks.,

**Comment by pklloveyou** [comment]
image

**Comment by pklloveyou** [comment]
It's really spectacular.

**Comment by Stoabrogga** [comment]
> 
> 51838,152166,152165,247237,132193,132194 creature_addon

The guids 152166,152165,247237,132193,132194 are not added because they have been added via creature_template_addon (entries 31333, 31334, 29747, 29790, 22441). Setis (GUID 51838) has been set to large via an additional statement, see the top of the SQL script:
```sql
-- Setis
UPDATE `creature_addon` SET `isLarge` = 1 WHERE `guid` = 51838;
```

**Comment by pklloveyou** [comment]
I'm trying one at a time.

**Comment by FrancescoBorzi** [comment]
@pklloveyou I don't think we need to check them all one by one :)

We really appreciate your help!

**Comment by Stoabrogga** [comment]
@pklloveyou 
Could you please tell me what you are testing? It is not necessary to test all the creatures, just a few of them to check if everything works. I wrote the SQL script in a way to not delete or remove anything, only add the large creature flags. This can also be easily reverted in the future if necessary.

**Comment by Stoabrogga** [comment]
I forgot a few entries for creature_template_addon and pushed a new commit. Should now be correct.

**Comment by pklloveyou** [comment]
Tested, beautiful.

**Comment by Knindzagxg** [comment]
Range working pretty good for me.
Tested in Hellfire Pen. and Nexus zone with GM on to see Maly, and had no issues at all.

**Comment by Stoabrogga** [comment]
Thanks for testing @pklloveyou @Knindzagxg 
I pushed another commit in order to set the 6 "Bone Giant" skeletons in Icecrown (The Valley of Lost Hope) to large. Don't think it is necessary to test these 6 creatures, so this PR should be ready to be merged.

**Comment by Mitradis** [comment]
if i set Visibility.Distance.Continents=250 these improvements will continue to work? Or with Visibility.Distance.Continents=250 this feature will be disabled?

**Comment by LichKing255** [comment]
excellent

**Comment by LichKing255** [comment]
> @pklloveyou
> Good idea, will check your list later. Some of the creatures (like the Fel Reavers) are already active and so have increased visibility (see their SAI), but they only will be triggered if a player is in normal range. By setting them to "large" they should be triggered at the full distance, which is better I think.

if PreloadAllNonInstancedMapGrids enabled , work.

**Comment by LichKing255** [comment]
@Stoabrogga 
A problem, If you're standing in a fixed place, you'll see it as Creature Large moves and goes away.
You will see it until you don't move. But then, if you go a little further, will no longer see creature and hide.

**Comment by wizzymore** [comment]
> @Stoabrogga
> A problem, If you're standing in a fixed place, you'll see it as Creature Large moves and goes away.
> You will see it until you don't move. But then, if you go a little further, will no longer see creature and hide.

I think that he wants to say:
If you stay in one place, you can see the Large Creature moving and going away ( i think out of the range where it should not be visible ) and if you move, when starting to move it will update and the creature will disappear. Tried to test this and can't confirm it as it didn't happen to me, but i don't think i got the message right

**Comment by Stoabrogga** [comment]
@LichKing255 
Yes, I noticed this kind of behaviour some time ago for active creatures, but it is not connected to this PR. I think that this is caused by the client.

**Comment by LichKing255** [comment]
> > @Stoabrogga
> > A problem, If you're standing in a fixed place, you'll see it as Creature Large moves and goes away.
> > You will see it until you don't move. But then, if you go a little further, will no longer see creature and hide.
> 
> I think that he wants to say:
> If you stay in one place, you can see the Large Creature moving and going away ( i think out of the range where it should not be visible ) and if you move, when starting to move it will update and the creature will disappear. Tried to test this and can't confirm it as it didn't happen to me, but i don't think i got the message right

It is true


> @LichKing255
> Yes, I noticed this kind of behaviour some time ago for active creatures, but it is not connected to this PR. I think that this is caused by the client.

But i changed to max visibility in setting game.

I can send video for this and different this with Warmane.

**Comment by Stoabrogga** [comment]
@LichKing255 
Then please open an issue for this, as this is not related to this PR. The issue exists also without this PR.

**Comment by LichKing255** [comment]
> @LichKing255
> Then please open an issue for this, as this is not related to this PR. The issue exists also without this PR.

But related with Large creatures. Wait for record video and see.

**Comment by Stoabrogga** [comment]
@LichKing255 
I know what you mean, no need for videos. But I cannot fix this in this PR because this PR is only about increasing drawing distance for large creatures and handling them like active ones if a player is near. The issue you describe is also true for active creatures which already have increased visibility and has to be solved in another PR.

**Comment by LichKing255** [comment]
> @LichKing255
> I know what you mean, no need for videos. But I cannot fix this in this PR because this PR is only about increasing drawing distance for large creatures and handling them like active ones if a player is near. The issue you describe is also true for active creatures which already have increased visibility and has to be solved in another PR.

I don't know what to write because my English is poor. If possible, create and explain a new Issue.
You can also use the video below in your Issue:
https://mega.nz/#!jtsjwSJJ!ImcWya0kzWBI2ZLdhKYIav83zwBWbyujWH3MSpCq89A

**Comment by Stoabrogga** [comment]
Ok, I'll create an issue for this. I noted this behaviour the first time while implementing the race at the Mirage Raceway, but though not much about it.

**Comment by Stoabrogga** [comment]
@LichKing255 
I created issue #2323 for this.

**Comment by LichKing255** [comment]
> @LichKing255
> I created issue #2323 for this.

TY

File: data/sql/updates/pending_db_world/rev_1569250786242020759.sql (hunk 1 above)
Comment (BarbzYHOOL) [reply data/sql/updates/pending_db_world/rev_1569250786242020759.sql]: yeah, so confusing, i remember reading this on google like 4 times xD

**Pull request merged**